File: data-structures/ds-cque/java/CircularQueue.java
public class CircularQueue
{
  private int SIZE = 5;
  private int front;
  private int rear;
  private int items[] = new int[SIZE];

    CircularQueue ()
  {
    front = -1;
    rear = -1;
  }

  // Check if the queue is full
  private boolean isFull ()
  {
    if (front == 0 && rear == SIZE - 1)
      {
	return true;
      }
    return front == rear + 1;
  }

  // Check if the queue is empty
  private boolean isEmpty ()
  {
    return front == -1;
  }

  // Adding an element
  private void enQueue (int element)
  {
    if (isFull ())
      {
	System.out.println ("Queue is full");
      }
    else
      {
	if (front == -1)
	  front = 0;
	rear = (rear + 1) % SIZE;
	items[rear] = element;
	System.out.println ("Inserted " + element);
      }
  }

  // Removing an element
  private int deQueue ()
  {
    int element;
    if (isEmpty ())
      {
	System.out.println ("Queue is empty");
	return (-1);
      }
    else
      {
	element = items[front];
	if (front == rear)
	  {
	    front = -1;
	    rear = -1;
	  }
	else
	  {
	    front = (front + 1) % SIZE;
	  }
	return (element);
      }
  }

  private void display ()
  {
    int i;
    if (isEmpty ())
      {
	System.out.println ("Empty Queue");
      }
    else
      {
	System.out.println ("Front -> " + front);
	System.out.println ("Items -> ");
	for (i = front; i != rear; i = (i + 1) % SIZE)
	  System.out.print (items[i] + " ");
	System.out.println (items[i]);
	System.out.println ("Rear -> " + rear);
      }
  }

  public static void main (String[]args)
  {

    CircularQueue q = new CircularQueue ();
    q.deQueue ();

    q.enQueue (1);
    q.enQueue (2);
    q.enQueue (3);
    q.enQueue (4);
    q.enQueue (5);
    q.enQueue (6);

    q.display ();

    int elem = q.deQueue ();

    if (elem != -1)
      {
	System.out.println ("Deleted Element is " + elem);
      }
    q.display ();

    q.enQueue (7);

    q.display ();

    q.enQueue (8);
  }

}

File: data-structures/ds-cque/javascript/circular_queue.js

var CircularQueue = function(size) {
    this.items = new Array(size)
    this.size = size
    this.length = 0

    this.front = size - 1
    this.rear = 0
};

// Insert an element in to the circular queue.
CircularQueue.prototype.enQueue = function(element) {
    if (this.isFull()){
        return false
    }else{
        this.items[this.rear] = element           
        this.rear = (this.rear + 1) % this.size
        this.length++
        return true
    }
};

// Extract the element at the front of the circular queue.
CircularQueue.prototype.deQueue = function() {
    if (this.isEmpty()){
        return false
    }else{

        const element = this.items[this.front % this.size]
        this.front = (this.front + 1) % this.size
        this.items[this.front % this.size] = null
        
        this.length--

        return element
    }
};

//Checks if the circular queue is empty.
CircularQueue.prototype.isEmpty = function() {
    return this.length === 0
};

//Checks if the circular queue is full.
CircularQueue.prototype.isFull = function() {
   return this.length == this.size
};

CircularQueue.prototype.display = function(){

    if(this.isEmpty()){
        console.log("The circular queue is empty.")
    }else{
        console.log("Front -> " + this.front)
        console.log("Items -> " + this.items.toString())
        console.log("rear -> " + this.rear)
    }
};

function main () {

    const cq = new CircularQueue(6)

    cq.display()
    console.log("")

    // We fill the queue.
    cq.enQueue(1)
    cq.enQueue(2)
    cq.enQueue(3)
    cq.enQueue(4)
    cq.enQueue(5)
    cq.enQueue(6)

    cq.display()
    console.log("")

    // We try to introduce another element, because it is full, it doesn't allow us.
    cq.enQueue(7)

    cq.display()
    console.log("")

    // we empty the queue.
    cq.deQueue()
    cq.deQueue()
    cq.deQueue()
    cq.deQueue()
    cq.deQueue()
    cq.deQueue()

    cq.display()
    console.log("")
}


main()

